### ISSUE TITLE
Cannot read property rules of undefined

### ISSUE THREAD

**Issue by vtereshyn** [cssinjs/jss#1188]
Hi. Not sure how to reproduce locally or create a code snippet with a reproducible issue, but got this today 5 times on stage environment. It's a really weird and very important issue since I can't debug it
Hope you can help. 

Thank you in advance

![image](https://user-images.githubusercontent.com/32247411/64632869-a54fbc00-d402-11e9-99a4-86b24e5ea01a.png)

**Comment by vtereshyn** [comment]
https://github.com/cssinjs/jss/pull/1189

**Comment by vtereshyn** [comment]
https://github.com/cssinjs/jss/pull/1189

**Comment by vtereshyn** [comment]
So, this issue brokes an application in production. Of course its my bad to use package alpha version in production, but here is a result.
React-jss is not ready for use in prod builds

**Comment by HenriBeck** [comment]
Could you please include more info about your setup and under which conditions you are using `react-jss` so we can look into this?

**Comment by vtereshyn** [comment]
@HenriBeck I couldn't create a codesandbox with a reproducible example, but I can try to describe:

I have a component with dynamic styles. This is FunctionComponent which uses `createUseStyles`. There I have a few CSS properties which dynamically change depends of Component props. 
I use this component on several pages. So when I return from one of the pages to another I got the error described above. In the console, I saw that the problem in `onUpdate` method or smth like that.

Also, before I had an issue with dynamical styles. My styles didn't update properly. So I needed a full page reload to update CSS. This is weird. 

But I'm looking forward to trying to help you in fixing these issues. 

Now I resolved these issue using different classnames depends on Component props.

**Comment by HenriBeck** [comment]
The not updating of styles should be fixed in #1190

**Comment by vtereshyn** [comment]
Cool, I'll update my jss versions and check. Thanks for the update.
Maybe my problem relates to #1190

**Comment by HenriBeck** [comment]
We haven't released a version with that fix yet

**Comment by vtereshyn** [comment]
@HenriBeck yes, I know. So I'm waiting for the new version and then update

**Comment by felthy** [comment]
I've been having this problem too. It won't be fixed by #1190, it's a different problem.

The crash can be triggered by using a dynamic style inside a media query e.g.:

```javascript
  button: {
    '@media (min-width: 0px)': {
      color: () => 'red'
    }
  }
```

The crash occurs if you unmount a component that uses this sheet, while there are still other instances mounted.

Here's a codesandbox:

https://codesandbox.io/s/jss-mq-dynamic-style-remove-ju7wr

Just click "Remove" to crash it.

@HenriBeck I think the problem here is that the `key` on `ConditionalRule` is not unique for per mounted instance, so when it is removed by `removeDynamicRules()` there is no longer a rule in `RuleList.map` with the key `@media (min-width: 0px)` so a subsequent call to `updateDynamicRules()` with that key will crash. Other rules e.g. `StyleRule` have unique keys e.g. `button-0`.

**Comment by HenriBeck** [comment]
> Also, before I had an issue with dynamical styles. My styles didn't update properly. So I needed a full page reload to update CSS. This is weird.

I was linking your fix to this problem. I'm aware that this problem is completely unrelated to yours

**Comment by felthy** [comment]
Yes but to clarify, my previous comment is all about this issue. The codesandbox above is new - I created it specifically for this issue because there was no sandbox yet to reproduce this bug.

**Comment by geraudi** [comment]
Hi
I have the same issue (Cannot read property rules of undefined)
This change fixed my problem :
```js
const useStyles = createUseStyles(theme => ({
    container: {
        padding: props => props.paddingContainer.xxs,
        [theme.breaks.xs]: {
            padding: props => props.paddingContainer.xs
        },
    },
   ```
by 
```js
const useStyles = createUseStyles(theme => ({
    container: props => ({
        padding: props.paddingContainer.xxs,
        [theme.breaks.xs]: {
            padding: props.paddingContainer.xs
        },
    }),
```

**Comment by iamstarkov** [comment]
it affects us too

**Comment by iamstarkov** [comment]
https://github.com/nordnet/nordnet-ui-kit/pull/649

**Comment by stoicsleuth** [comment]
Is there any update on this issue?

**Comment by alex-shamshurin** [comment]
+

**Comment by hartbeatnt** [comment]
This is affecting my company's production site as well, any timeline on when a version with the fix will be released?

**Comment by iamstarkov** [comment]
@hartbeatnt you can submit a PR with a fix and then it will be released shortly after successful review

**Comment by alex-shamshurin** [comment]
Any update? Does anybody knows where is it ?

**Comment by kof** [comment]
Seem like a bad bug, does anyone want to create a test or/and a fix?

**Comment by iamstarkov** [comment]
@kof this can be used for a test https://github.com/nordnet/nordnet-ui-kit/pull/649/files

**Comment by vladanyes** [comment]
Waiting for the new version too

**Comment by hi-zhaolei** [comment]
I think I found the reason for this problem. 

We met this issue by use react-hot-loader. When some component code change. webpack retranspile and make hot reload. Then the component use jss hook will show this error and will make our program broken down. It makes me really headache.

After some debug. I found the parameter rule is undefined because the property dynamicRules on state not rewritten when state.sheet removes the rules.

For details. 

In the picture following.  We use addDynamicRules to add rulelist on the sheet and assigned to state.dynamicRules. 

![image](https://user-images.githubusercontent.com/5736062/70851337-e4445080-1ece-11ea-8a26-6ee298a32f82.png)

And in effect clean up callback. Only remove all rules but forgot to reset property dynamicRules on state

![image](https://user-images.githubusercontent.com/5736062/70851342-ed352200-1ece-11ea-9cb4-eda81d998f2a.png)

So. when the next effect hook execute. We can`t found the dynamicRules in rulelist because it`s already removed before.

I'm not sure the code above is intentional or it`s a bug. Hope someone can response.
